### PR TITLE
HPCC-29053 Add --init-publisher-wuid option to ecl command line

### PR DESCRIPTION
**Pull request by afishbeck**
Add --init-publisher-wuid option wherever the cli supports --dfu-copy.  --init-publisher-wuid will preallocate the publisher wuid and display it on the command line before running the rest of the command.  This will allow the user to track the progress even if the cli interface gets disconnected from the server before the command completes.

This commit also improves the setting of the Publisher Workunit job name making it easier to find and understand the context of publisher workunits.

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: ecl/ecl-bundle/CMakeLists.txt [hunk 1]
```diff
@@ -54,9 +55,10 @@ include_directories (
          ${HPCC_SOURCE_DIR}/ecl/eclcmd
          ${HPCC_SOURCE_DIR}/ecl/eclcc
          ${HPCC_SOURCE_DIR}/common/thorhelper
+         ${ESPSCM_GENERATED_DIR}
     )
 
-ADD_DEFINITIONS( -D_CONSOLE -DWSSMC_API_LOCAL )
+ADD_DEFINITIONS( -D_CONSOLE -DWSSMC_API_LOCAL -DFileSpray_API_LOCAL )
```
Comment (ghalliday) [inline ecl/ecl-bundle/CMakeLists.txt:61]: This suggests the existing structure isn't quite right.  Why would the ecl-bundle command have a dependency on the spray interface?

File: ecl/eclcmd/eclcmd_common.cpp [hunk 2]
```diff
@@ -895,3 +896,38 @@ eclCmdOptionMatchIndicator EclCmdWithQueryTarget::parseCommandLineOptions(ArgvIt
...
+    catch (IException *E)
+    {
+        //if we can't preallocate the publisher DFU workunit it might be because the server is too old, an expected scenario pre/early release 9.x, we'll just create the publisher wuid the old way
+        StringBuffer msg;
+        int code = E->errorCode();
+        E->errorMessage(msg);
+        E->Release();
+
+        //Not ideal way of detecting this, but all we're trying to do is to improve the info message when the server is old
+        if (code == -2 && strieq(msg, "400: Bad Request [Method CreateDFUPublisherWorkunit not available in service FileSpray"))
```
Comment (ghalliday) [inline ecl/eclcmd/eclcmd_common.cpp:918]: This looks like it is missing a closing ']', but that may be correct.  Looking at the code I suspect not.

Possibly better would be

strstr(msg, "Method CreateDFUPublisherWorkunit not available")

File: ecl/eclcmd/eclcmd_common.hpp [hunk 3]
```diff
@@ -499,16 +513,20 @@ class EclCmdOptionsDFU
             "   --dfu-overwrite        Set DFU copy command to overwrite physical files that are already on disk.\n"
             "   --only-copy-files      Copy the files needed for the query, but don't publish the query\n"
             "   --stop-if-files-copied If all files already exist, publish the query.\n"
-            "                          Otherwise, copy the files needed for the query, but don't publish the query\n",
+            "                          Otherwise, copy the files needed for the query, but don't publish the query\n"
+            "   --init-publisher-wuid  Allocate and display the publisher wuid immediately, so that it can be tracked\n"
+            "                           even if the command line is diconnected\n",
```
Comment (ghalliday) [inline ecl/eclcmd/eclcmd_common.hpp:518]: typo: disconnected

**Review by ghalliday — COMMENTED** [review]
@afishbeck a couple of comments.  I don't expect you to think about restructuring for this PR.
Is the target appropriate?  Is it a feature that is needed now, or should it wait until 9.2?  (open question).

File: ecl/ecl-bundle/CMakeLists.txt (hunk 1 above)
Comment (afishbeck) [reply ecl/ecl-bundle/CMakeLists.txt:61]: Agreed.  Need to look at when/how EclCmdOptionsDFU is incorporated.  

I've opened JIRA https://track.hpccsystems.com/browse/HPCC-29460

File: ecl/eclcmd/eclcmd_common.cpp (hunk 2 above)
Comment (afishbeck) [reply ecl/eclcmd/eclcmd_common.cpp:918]: The string is correct, the closing ']' is missing I did test this.

But agreed, I'll switch to a strstr.

**Comment by afishbeck** [comment]
> @afishbeck a couple of comments. I don't expect you to think about restructuring for this PR. Is the target appropriate? Is it a feature that is needed now, or should it wait until 9.2? (open question).

This is one of the few remaining issues related to the adoption of the --dfu-copy option.  The other two are actually Jake's.  I'm hoping to get these into their hands, but can retarget to master if you want.  Especially since they still need Jake's changes.

**Comment by afishbeck** [comment]
@ghalliday some minor updates and rebased to master.

**Comment by afishbeck** [comment]
@ghalliday FYI the insurance cloud team is keen on this one.

**Review by ghalliday — APPROVED** [review]
Not sure which version to target though....

**Pull request merged**